mjpc/planners/sample_gradient/planner.cc

```diff
@@ -0,0 +1,652 @@
+      }
+
+      // ----- rollout sample policy ----- //
+
+      // policy
+      auto sample_policy_i = [&candidate_policy = s.candidate_policy, &i](
+                                 double* action, const double* state,
+                                 double time) {
+        candidate_policy[i].Action(action, state, time);
+      };
+
+      // policy rollout
+      s.trajectory[i].Rollout(
+          sample_policy_i, task, model, s.data_[ThreadPool::WorkerId()].get(),
+          state.data(), time, mocap.data(), userdata.data(), horizon);
+    });
+  }
+  pool.WaitCount(count_before + num_trajectory);
+  pool.ResetCount();
+}
+
+// compute candidate trajectories
+void SampleGradientPlanner::GradientCandidates(int num_trajectory,
+                                               int num_gradient, int horizon,
+                                               ThreadPool& pool) {
+  if (num_gradient < 1) return;
+
+  // number of parameters
+  int num_parameters = resampled_policy.num_parameters;
+  int num_spline_points = resampled_policy.num_spline_points;
+
+  // cache old gradient
+  mju_copy(gradient_previous.data(), gradient.data(), num_parameters);
+
+  // -- compute approximate gradient -- //
+  // average return
+  int num_noisy = num_trajectory - num_gradient;
+  double avg_return = 0.0;
+  for (int i = 0; i < num_noisy; i++) {
+    avg_return += trajectory[i].total_return;
+  }
+  avg_return /= num_noisy;
+
+  // gradient
+  std::fill(gradient.begin(), gradient.end(), 0.0);
+  for (int i = 1; i < num_noisy; i++) {
+    double* noisei = noise.data() + i * (model->nu * kMaxTrajectoryHorizon);
+    mju_addToScl(gradient.data(), noisei,
+                 (trajectory[i].total_return - avg_return) / (num_noisy - 1),
+                 num_parameters);
+  }
+
+  // normalize gradient
+  // TODO(taylor): should we normalize?
+  mju_normalize(gradient.data(), num_parameters);
+
+  // compute step sizes along gradient
+  std::vector<double> step_size(num_gradient);
+  LogScale(step_size.data(), gradient_max_step_size, gradient_min_step_size,
+           num_gradient);
+
+  // gradient filter gf * grad + (1 - gf) * grad_prev
+  double gradient_filter = gradient_filter_;
+
+  // compute candidate policies along gradient direction
+  // these candidates will be evaluated at the next planning iteration
+  for (int i = num_noisy; i < num_trajectory; i++) {
+    // copy nominal policy
+    candidate_policy[i].CopyFrom(resampled_policy, num_spline_points);
+    candidate_policy[i].representation = resampled_policy.representation;
+
+    // scaling
+    // TODO(taylor): scale by num_parameters?
+    // TODO(taylor): divide by reward std dev?
+    double scaling = step_size[i - num_noisy] / noise_exploration;
+
+    // gradient step
+    mju_addToScl(candidate_policy[i].parameters.data(), gradient.data(),
+                 -scaling * gradient_filter, num_parameters);
+
+    // TODO(taylor): resample the gradient_previous?
```

Comment:
Ya maybe. @nimrod-gileadi's thinking, which makes sense to me, is that instead of saving gradient and gradient_previous, it would be something like gradient_proposal (scaling==1) and winner_previous. So instead of propagating gradients for which resampling is not well-defined, we propagate candidates. However with the upcoming Sliding Splines the "resampling" procedure is a no-op except for iterations where the oldest point is erased and a new point is invented, which would change the logic anyway.